zerver/lib/request.py

```diff
@@ -24,6 +27,43 @@
 
 from zerver.lib.exceptions import ErrorCode, InvalidJSONError, JsonableError
 from zerver.lib.types import Validator, ViewFuncT
+from zerver.models import Client, Realm
+
+
+@dataclass
+class ZulipRequestData:
+    """This contains the additional attributes for a request that we
+    add in zerver/lib/middleware.py. Presumably, certain fields are
+    expected to not None in `request_data_map[request]` as long as
+    the request is being accessed after being processed in the middleware.
+
+    """
+
+    client: Optional[Client] = None
+    client_name: Optional[str] = None
+    client_version: Optional[str] = None
+    _log_data: Optional[MutableMapping[str, Any]] = None
+    _rate_limit: Optional[str] = None
+    _requestor_for_logs: Optional[str] = None
+    realm: Optional[Realm] = None
```

Comment:
This one is tricky, since we currently distinguish between `request.realm` being absent and being `None`.

https://github.com/zulip/zulip/blob/e91054ef1935c339f154937bc619c72c3bfbd33d/zerver/context_processors.py#L55-L65

Reply:
Hmm, how about having a `realm_cached` flag?

Reply:
Maybe `has_fetched_realm` would be better?  

I think also we should be setting that parameter in a `finally` clause of the middleware, since in all cases we've attempted to fetch the realm and determined whether there is one or not associated with the URL in question.

Reply:
A `finally` clause of what? If this throws an exception (other than the `Realm.DoesNotExist` exception this already catches), the request is going to fail anyway.

Reply:
What I meant is that the middleware could provide the guarantee that `request_notes.realm` is always set after middleware (to an `Optional[Realm]`), but currently doesn't.  And that if we did that, we might get to delete `request_notes.has_fetched_realm`.